### PR TITLE
RepeatingTaskThread

### PR DESCRIPTION
**Pull request by TomaszTB**
Replacement for `RestartableThread` and `RestartableThrottledThread`!

### CODE REVIEW

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+   /**
+    * Whether this thread is running. In other words, whether this thread has not been {@link #kill()}ed.
+    * <p>
+    * The returned value of this method does not necessarily equate to {@link #isAlive()},
+    * as the loop may take some time to finish after the call to {@link #kill()},
+    * during which the thread remains alive.
+    *
+    * @return {@code false} if {@link #kill()} or {@link #blockingKill()} has been called. {@code true} otherwise.
+    */
+   public boolean isRunning()
```
Comment (TomaszTB) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:200]: This method is extremely similar to `isAlive()`. Should it be kept?

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+    * <p>
+    * You may {@code @Override} this method with the code to repeat.
+    * Alternatively, if a {@link RunnableThatThrows} was passed it, this method
+    * will call the run method in the loop.
+    *
+    * @throws Throwable Any throwable that the overriding code or the passed in task may throw.
+    *       This throwable will be handled by the passed in {@link ExceptionHandler}
+    *       (by default it is {@link DefaultExceptionHandler#MESSAGE_AND_STACKTRACE}).
+    */
+   protected void repeat() throws Throwable
```
Comment (TomaszTB) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: Changed from `runInLoop()` to `repeat()`. What do we think?

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 1 above)
Comment (calvertdw) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:200]: Would you use it anywhere?

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 2 above)
Comment (calvertdw) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: Why not runTask()?

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 3]
```diff
@@ -0,0 +1,390 @@
+package us.ihmc.commons.thread;
+
+import us.ihmc.commons.Conversions;
+import us.ihmc.commons.RunnableThatThrows;
+import us.ihmc.commons.exception.DefaultExceptionHandler;
+import us.ihmc.commons.exception.ExceptionHandler;
+import us.ihmc.commons.exception.ExceptionTools;
+
+/**
+ * A thread that executes code in a loop.
```
Comment (calvertdw) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:10]: A thread that repeats execution of a single task. It can do this N times, continuously, or at a constant rate.

Comment (calvertdw) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]:
```suggestion
 * A thread that repeats execution of a single task. It can do this N times, continuously, or at a constant rate.
```

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+   /**
+    * Countdown for number of repetitions to run.
+    * The counter is decremented each time before the repetition.
+    * Once the counter hits 0, the loop is paused.
+    * <ul>
+    *    <li> 0 = pause (don't run the loop until counter value is changed).
+    *    <li> -1 = repeat indefinitely (keep looping until told otherwise).
+    *    <li> N > 0 = run the loop N more repetitions.
+    */
+   private volatile int remainingRepetitions = 0;
```
Comment (calvertdw) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: Should this be a long too?

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 1 above)
Comment (TomaszTB) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:200]: Nope, I can't think of a place where I'd use it

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 2 above)
Comment (TomaszTB) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: That sounds pretty good

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 4 above)
Comment (TomaszTB) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: Oh yeah, that'd be good

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+
+   /** Counter for the total number of repetitions completed during the lifetime of this thread. */
+   private long completedRepetitions = 0L;
+
+   /**
+    * Indicates whether this object is destroyed.
+    * The loop will come to a finish when {@code isRunning() == false}.
+    * Does not equate to {@link Thread#isAlive()}, as the thread may take
+    * some time to finish executing after {@code isRunning()} becomes {@code false}.
```
Comment (calvertdw) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: I would try and reword this to encapsulate the following, or just replace with the following. I think this variable doesn't have a super meaningful meaning and isn't part of the API so doesn't need a meaningful description.

Initially false. When true, the Thread has been started. When true -> false, the main loop allows the currently executing task (if any) to complete and then allows the thread to die.

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+   }
+
+   public RepeatingTaskThread(ExceptionHandler exceptionHandler, String name)
+   {
+      this(exceptionHandler, UNLIMITED_FREQUENCY, name);
+   }
+
+   public RepeatingTaskThread(ExceptionHandler exceptionHandler, double loopFrequencyLimit, String name)
+   {
+      this(null, exceptionHandler, loopFrequencyLimit, name);
```
Comment (calvertdw) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: What's the point of these constructors that have a null task?

Comment (TomaszTB) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: They're to be used when you override the `runTask()` method instead of passing in a task

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+   {
+      this(task, DefaultExceptionHandler.MESSAGE_AND_STACKTRACE, loopFrequencyLimit, name);
+   }
+
+   public RepeatingTaskThread(RunnableThatThrows task, ExceptionHandler exceptionHandler, String name)
+   {
+      this(task, exceptionHandler, UNLIMITED_FREQUENCY, name);
+   }
+
+   public RepeatingTaskThread(RunnableThatThrows task, ExceptionHandler exceptionHandler, double loopFrequencyLimit, String name)
```
Comment (calvertdw) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: I would try to remove loopFrequencyLimit from the constructor. You could make setFrequencyLimit return this to allow fluent-style construction.

Comment (calvertdw) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: Sometimes fluent construction would use withFrequencyLimit for that

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+   {
+      super(name);
+      this.task = task;
```
Comment (calvertdw) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:79]: Actually here I think you should do `this.task = ExceptionTools.handle(task, exceptionHandler);`
and remove exceptionHandler field and make task field just type Runnable.

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+               throttler.waitAndRun(loopPeriodLowerLimit);
+            }
+         }
+         catch (InterruptedException interrupted)
+         {  // Maintain interrupted status so that runInLoop can handle it
+            interrupt();
+         }
+
+         // Run the runInLoop method, and handle any exception it may throw.
+         ExceptionTools.handle(this::repeat, exceptionHandler);
```
Comment (calvertdw) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: This should just be `task.run()` where the exception handling is built in the constructor. Then you can delete the repeat/runTask method.

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 1 above)
Comment (calvertdw) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:200]: `isRunning()` and `isRepeating` are not intuitive. I would just provide `getCompleted` and `getRemaining`

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 9 above)
Comment (TomaszTB) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: The `runTask` method is needed so it can be overridden, but it could just call `runTask()`

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+         ++completedRepetitions;
```
Comment (calvertdw) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:392]: What if you decrement remaining down here, so then you don't have a moment where remaining == 0 but it's working on executing the task.

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+    * <p>
+    * This method does not do anything if the thread is repeating indefinitely.
+    *
+    * @param repetitions The number of repetitions to add. This can be a negative value for subtraction.
+    */
+   public void addRemaining(int repetitions)
+   {
+      synchronized (loopLock)
+      {
+         // If looping indefinitely, do nothing
```
Comment (calvertdw) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]:
```suggestion
         // If repeating indefinitely, do nothing
```

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+            remainingRepetitions = 0;
+
+         loopLock.notify();
+      }
+   }
+
+   /**
+    * Get the remaining number of repetitions this thread plans to run.
+    *
+    * @return The remaining number of loops this thread plans to run.
```
Comment (calvertdw) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]:
```suggestion
    * @return The remaining number of repetitions to execute.
```

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 1 above)
Comment (TomaszTB) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:200]: I can remove `isRunning`, but I'll probably keep `isRepeating` package private to use in tests. It is handy there

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 7 above)
Comment (TomaszTB) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: I think it should be kept as "set" since it can be used at anytime, not only during construction

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 9 above)
Comment (calvertdw) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: Okay that sounds good

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 8 above)
Comment (calvertdw) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:79]: Nevermind, didn't realize overriding runTask was part of the API.

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 10 above)
Comment (TomaszTB) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:392]: When the task is being executed the repetition has begun, so I'd say 0 remain to be executed. At least that's how I've been thinking about it. Also it's nice to keep it in one `synchronized` block haha. I'm still happy to make this change if you insist

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+   }
+
+   /**
+    * Signals the thread to start repeating indefinitely.
+    * If the thread has not been started yet, calling this
+    * method will also start the thread.
+    */
+   public void startRepeating()
+   {
+      if (!isAlive())
```
Comment (calvertdw) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]:
```suggestion
      if (!running)
```

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 5 above)
Comment (TomaszTB) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: I changed the comment. Let me know if it's good

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 10 above)
Comment (calvertdw) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:392]: Currently there is no way to tell if a task is currently executing or not. It would be great is repetitions have 3 states SCHEDULED -> EXECUTING -> COMPLETED and that be thread safe state. I would advocate adding a boolean for isExecuting, but also, I think the "remaining" term is ambiguous for whether or not it's remaining _while_ it's executing.

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 14]
```diff
@@ -0,0 +1,489 @@
...
+   {
+      state.addRemaining(repetitions);
+   }
+
+   /**
+    * Access the state of this thread.
+    *
+    * @return The state of this thread.
+    */
+   public State getRepetitionState()
```
Comment (calvertdw) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: I think the user should not even know about the internal State class or interact with it.

Comment (TomaszTB) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: Done!

**Review by ds58 — APPROVED** [review]

**Review by calvertdw — APPROVED** [review]

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 15]
```diff
@@ -343,17 +343,14 @@ public final void run()
...
          if (periodLowerLimit > 0.0)
          {
-            /*
-             * This call must not swallow interrupts.
-             * As of writing this comment (Nov, 2024), LockSupport.parkNanos() is used internally to block.
-             * Although the throttler will block until the period has elapsed, the thread
-             * remains interrupted.
-             */
+            // Uses ThreadTools.parkAtLeast, which will not throw InterruptedException, but
+            // will return early when this thread is interrupted.
```
Comment (TomaszTB) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:344]: I think this is false, because parkAtLeast loops until the desired time has elapsed. It might just free spin if parkNanos is returning immediately due to the interrupt. We should probably fix that haha.

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 16]
```diff
@@ -343,17 +343,14 @@ public final void run()
...
             throttler.waitAndRun(periodLowerLimit);
-         }
 
-         // clear interrupted status
-         interrupted();
+            // clears interrupted status from
+            if (interrupted() && executionState.getScheduled() == 0L)
```
Comment (TomaszTB) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: I agree with this behavior. But what about clearing the interrupt status somewhere else, so it doesn't depend on whether the thread is throttled?

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 15 above)
Comment (calvertdw) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:344]: I fixed this in https://github.com/ihmcrobotics/ihmc-commons/pull/41

**Review by calvertdw — APPROVED** [review]

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 17]
```diff
@@ -0,0 +1,448 @@
...
+   {
+      this(null, exceptionHandler, name);
+   }
+
+   public RepeatingTaskThread(RunnableThatThrows task, String name)
+   {
+      this(task, DefaultExceptionHandler.MESSAGE_AND_STACKTRACE, name);
+   }
+
+   public RepeatingTaskThread(RunnableThatThrows task, ExceptionHandler exceptionHandler, String name)
```
Comment (ds58) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: `name` should come first in the parameters

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 18]
```diff
@@ -0,0 +1,448 @@
...
+         }
+      }
+
+      return false;
+   }
+
+   /**
+    * Allows the thread to die.
+    * Any currently executing task will first be allowed to finish.
+    * This class cannot be reused after this point.
```
Comment (ds58) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:272]: Is there a way to ensure it can't be reused?

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 19]
```diff
@@ -0,0 +1,448 @@
...
+
+         notifyAll();
+      }
+
+      /**
+       * Wait until a change occurs to the thread's execution state.
+       *
+       * @return {@code true} if interrupted.
+       */
+      private synchronized boolean waitForChange()
```
Comment (ds58) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:419]: Does this need to be synchronized? Seems weird to me

**Review by ds58 — CHANGES_REQUESTED** [review]

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 17 above)
Comment (TomaszTB) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: Dang, I was trying to keep the constructor similar to Thread, where name is after the runnable. Not a big deal though, and I also think name being first looks better in code haha

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java [hunk 20]
```diff
@@ -0,0 +1,448 @@
...
+   private volatile boolean running = false;
+   
+   /** Execution state of this thread */
+   private final ExecutionState executionState = new ExecutionState();
+
+   /** Throttler used to limit the execution frequency. */
+   private final Throttler throttler = new Throttler();
+
+   /** The period of the set frequency limit. A negative value indicates no limit. */
+   private volatile double periodLowerLimit = UNLIMITED_FREQUENCY;
```
Comment (ds58) [inline src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: I would rename this `frequencyLimit` like the setter method, it's more clear. Also, I don't see why it needs to be volatile?

**Review by ds58 — CHANGES_REQUESTED** [review]

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 19 above)
Comment (TomaszTB) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:419]: Yeah, the call to `wait()` requires it to be synchronized

Comment (ds58) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:419]: Oh, right. I forgot

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 20 above)
Comment (TomaszTB) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: Problem is that Throttler takes in the period for `waitAndRun()`, not frequency. I prefer thinking in hertz, so the setter is for a frequency limit. But it gets converted to seconds so we can use it with the throttler's method. 

Should we add a setter for period limit? 

I don't think it's crucial to be volatile though. I'll make that change

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 18 above)
Comment (TomaszTB) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:272]: Do you mean like throwing an exception if someone tries to restart it?

Comment (ds58) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:272]: Yes, that can work

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 20 above)
Comment (ds58) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: alright, I understand. I guess the naming can stay.

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 17 above)
Comment (calvertdw) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: In that case the task should always be last, but I did like the way Tomasz had it here

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 18 above)
Comment (calvertdw) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:272]: You'll already get one
image

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 20 above)
Comment (calvertdw) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: Agreed to make it not volatile

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 17 above)
Comment (ds58) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: Yes, I would prefer if `name` was first and `task` was last

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 18 above)
Comment (calvertdw) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:272]: Might be worth adding a test to assert this

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 17 above)
Comment (TomaszTB) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: I like having exception handler after runnable because you can do something like:
```
RepeatingTaskThread myThread = new RepeatingTaskThread("name", () -> 
{
   // Do stuff
}, exception -> 
{
   // Handle exceptions
});
```
It almost looks like a try catch block, which is very familiar to us. If exception handler came first it'd be a catch try block? :shrug:

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 18 above)
Comment (TomaszTB) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java:272]: Come to think of it, kill then start still works. Technically there's nothing wrong with that, except that it doesn't make sense

File: src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java (hunk 17 above)
Comment (ds58) [reply src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java]: Hmm, I suppose that's a decent reason to keep the current ordering

**Review by ds58 — APPROVED** [review]
